### ISSUE TITLE
unexpected ucrtbase.dll

### ISSUE THREAD

**Issue by P3t3rU5** [oneclick/rubyinstaller2#308]
## What problems are you experiencing?
`unexpected ucrtbase.dll` when running anything ruby

## Steps to reproduce
any ruby executable execution like `ruby -v`, `gem list`, etc

## What's the output from `ridk version`?
`unexpected ucrtbase.dll`

Furthermore, I'm on Windows 11 Pro Insider Preview build 25211.rs_prerelease.220923-1354 and I'm suspecting something changed in latest build

**Comment by larskanis** [comment]
OK, this is rather unexpected :smile:

Which RubyInstaller version did you use? Can you compare 3.0 and 3.1 versions? RubyInstaller-3.1 switched to more modern UCRT runtime, while 3.0 still uses the MSVCRT.

**Comment by P3t3rU5** [comment]
Didn't even thought of trying other versions
I had some previous versions installed:
* 3.0.1 verion is working
* 3.1.1 is not working
* 3.1.2 is not working

**Comment by s00ner** [comment]
I'm on the same Windows 11 version and also recently started getting `unexpected ucrtbase.dll` when running anything ruby.

**Comment by MSP-Greg** [comment]
@ArminG-MSFT Thank you.  Added your comments to https://bugs.ruby-lang.org/issues/18605#note-4

**Comment by gonzaloriestra** [comment]
@ArminG-MSFT thanks a lot for the detailed explanation. Could you be so kind to let us know once the fix is released?

**Comment by olifink** [comment]
@ArminG-MSFT - I just tried rubyinstaller-devkit-3.1.2-1-x64.exe on Window on Arm (Windows DevKit 2023) with Windows Pro 11 Insider Preview (dev channel) on 25236.rs_prerelease.221028-1618 and still get `unexpected ucrtbase.dll` trying to run ruby

**Comment by olifink** [comment]
thanks @ArminG-MSFT 25250 is now working for me on Windows on Arm Dev Kit 2023 device - thanks!

**Comment by nurse** [comment]
Hi,
I'm a person who introduced the `__pioinfo` hack, and notice this thread just now. Why Ruby depends to _pioinfo is

* to associate socket and fd: CRuby creates fd with dummy file handle and set socket to emulate Unix-like behavior
* to implement overlapped I/O for Windows 2000/XP
* to emulate fcntl(2)

I'm very sorry that Visual C++'s optimization is blocked by my hack. But unfortunately our decision to merge the change is not made as a tiny hack. After first I came up with the idea, we investigated other solution for a year. And also I believe many people including you struggled to find a smarter way, but no one proposed a solution.

I agree that Ruby should support ARM64 natively. I'm very welcome to contribute a patch to get `__pioinfo` on ARM64.

Of course I'm also welcome to propose a better solution.
Thanks,

**Comment by larskanis** [comment]
@nurse Thank you for sharing these details! I think it's best to move the discussion to bugs.ruby-lang.org where I [posted my answer](https://bugs.ruby-lang.org/issues/18605?next_issue_id=18603&prev_issue_id=18608#note-6). Maybe the title could be somewhat more generic there.

@ArminG-MSFT  Thank you for your investigation and your short term solutions! IMHO this particular issue with Windows 11 Insider Preview is solved that way.

**Comment by KJTsanaktsidis** [comment]
It took me a bit of fiddling to get the compatibility hack in Windows build 25250 because I'm not really a Windows person; in the hope it will be useful to anybody else trying to build Ruby from source, the way to get this fix is:

* Get on the Windows Insider Canary channel (NOT the dev channel - they swapped around what branch was dev/canary earlier this year - https://blogs.windows.com/windows-insider/2023/03/06/whats-coming-for-the-windows-insider-program-in-2023/

Or.... if you don't want to do that to your gaming PC (that's me!)

* Install the SDK for a build > 25250 from FlightHub - https://learn.microsoft.com/en-us/windows-insider/flight-hub/ (e.g. I downloaded & installed https://software-static.download.prss.microsoft.com/dbazure/Windows_InsiderPreview_SDK_en-us_25357_1.iso).

If you install the SDK, and then run "x64 native tools command prompt for Visual Studio 2022", it seems that it picks up the new SDK version - e.g. you'll see it in `%WindowsLibPath%` env var et al.

**Comment by KJTsanaktsidis** [comment]
Yes - I ran into this issue while building ruby (by hand in x64 VS console) on what I thought was an up to date win11 2H22 system yesterday. Building against a newer canary SDK fixed it for me.

I had assumed the patch just hadn’t made it out to windows update for me yet - are you saying it _is_ in the out in the normal release stream?

it’s of course possible I did something else to resolve it in my flailing about. There was a while where I was accidentally doing a 32 bit build which definitely won’t work based on what you’re saying for example.

Also just as a side note, this is a Herculean effort at backward comparability here by Microsoft, and I’m in awe of it.

**Comment by KJTsanaktsidis** [comment]
Ok I’m probably 90% sure I only got this error on the 32bit build on 2H22 then. Sorry for muddying the waters.

> But 22xxx builds should be unaffected. If they are not for 64bit I would want to know!

I don’t have any way to verify this without reinstalling windows on my computer now that I upgraded to the Dev track right? Is there a way I can build and run Ruby against an older version of the SDK perhaps and ensure ruby runs against that version of the DLL?

I’m starting to think what I did by installing the SDK was pointless  because ruby.exe will load the ucrtbase.DLL from C:\Windows\SysWOW64 at runtime anyway and ignore the DLL in the dev kit directory?

sorry for being a bit of a noob here.

**Comment by awhitenose** [comment]
Probably this is useful. I created a patch to win32.c which is based on win32.c taken from Ruby 3.2.x but really it may be applied and was tested with Ruby 2.6.x up to 3.2.x. It fixes 'unexpected ucrtbase.dll' issue for WIn32 builds and lets run 32-bit Ruby on Win11 as well as older Windows. 

Before making the patch, I have disassembled ucrtbase.dll from no-issues 32-bit windows, 64-bit, and problematic Win11 32-bit and 64-bit. Checked __pioinfo lookup with all the four by looking into asm and finding patterns. The only difference was using LEAVE in the newer 32-bit ucrtbase.dll vs POP EBP in the previous. The main MOV/LEA pattern was not changed and may be used as is. There are no changes in 64-bit asm that would prevent current patterns from working. I did not need 64-bit, but checked asm.

As a result this patch only adds the additional check for LEAVE instruction for 32-bit versions only. Win11 ucrtbase.dll _isatty is longer than 300 bytes, hence changed to 500 which helped.

```
--- win32.c.orig	2023-02-08 07:02:20.000000000 +0100
+++ win32.c	2023-06-24 00:58:44.000000000 +0100
@@ -2596,8 +2596,8 @@
      * * https://bugs.ruby-lang.org/issues/11118
      * * https://bugs.ruby-lang.org/issues/18605
      */
-    char *p = (char*)get_proc_address(UCRTBASE, "_isatty", NULL);
-    char *pend = p;
+    unsigned char *p = (char*)get_proc_address(UCRTBASE, "_isatty", NULL);
+    unsigned char *pend = p;
     /* _osfile(fh) & FDEV */
 
 # ifdef _WIN64
@@ -2622,10 +2622,16 @@
 #  define PIOINFO_MARK "\x8B\x04\x85"
 # endif
     if (p) {
-        for (pend += 10; pend < p + 300; pend++) {
+        for (pend += 10; pend < p + 500; pend++) {
             // find end of function
+#if _WIN64
             if (memcmp(pend, FUNCTION_BEFORE_RET_MARK, sizeof(FUNCTION_BEFORE_RET_MARK) - 1) == 0 &&
-                (*(pend + (sizeof(FUNCTION_BEFORE_RET_MARK) - 1) + FUNCTION_SKIP_BYTES) & FUNCTION_RET) == FUNCTION_RET) {
+                (*(pend + (sizeof(FUNCTION_BEFORE_RET_MARK) - 1) + FUNCTION_SKIP_BYTES) & FUNCTION_RET) == FUNCTION_RET) 
+#else
+      	    if ((*pend == 0x5d /*pop ebp*/ || *pend == 0xc9 /*leave*/) &&
+		*(pend+1) == 0xc3 /*ret*/)
+#endif	    
+	    {
                 // search backwards from end of function
                 for (pend -= (sizeof(PIOINFO_MARK) - 1); pend > p; pend--) {
                     if (memcmp(pend, PIOINFO_MARK, sizeof(PIOINFO_MARK) - 1) == 0) {

```

**Comment by awhitenose** [comment]
Thank you @ArminG-MSFT for your comments. Do you mean that the recent ucrtbase.dll must not cause issues with Ruby builds and running? This is good news if so, and then it's probably me but also some of the clients are still on previous builds for any reasons and that was why I decided to create the patch myself and finally get a working Ruby for us and them. 

Please note, the above patch is not applied to Ruby sources. I'm not one of Ruby developers or collaborators. However, if someone finds my fix useful, please feel free to apply it to Ruby. I think it still may be useful for someone who builds a custom Ruby from sources and finds it not working on some (yet not recent) versions of Windows.

**Comment by pbo-linaro** [comment]
I implemented `__pioinfo` pointer detection for (native) windows-arm64 in https://github.com/ruby/ruby/pull/8995.
You're welcome to join and review it.

**Comment by pbo-linaro** [comment]
@ArminG-MSFT If you could ensure `_isatty` in `ucrtbase.dll` for windows-arm64 is not modified too, that would be very welcome to help support this platform.

**Comment by pbo-linaro** [comment]
Only `adrp/add` pair is used for detection. If it gets broken, no worry, I'll implement a new monkey patch to support whatever code layout is generated.

To ensure it's the latest before ret, maybe you can simply introduce a [compiler barrier](https://learn.microsoft.com/en-us/cpp/intrinsics/readwritebarrier?view=msvc-170) there, so no other variable will be loaded instead of `__pioinfo`.

Out of curiosity, do you know why `__pioinfo` public visibility was removed? Today, it seems like MS had to ensure compatibility based on code layout (as you implemented), when it could be much easier by simply exporting the symbol, like it was before.

**Comment by pbo-linaro** [comment]
Thank you for the great insight on this, it's interesting to read this from an *internal* developer.

> Last, note that the dummy assembly I'm adding is never executed. So there is no need for a barrier. Ruby is only passively scanning it.

My point on using a [`_ReadWriteBarrier`](https://learn.microsoft.com/en-us/cpp/intrinsics/readwritebarrier?view=msvc-170) (or as documentation suggests, [`atomic_thread_fence`](https://learn.microsoft.com/en-us/cpp/standard-library/atomic-functions?view=msvc-170#atomic_thread_fence)) is that it will prevent compiler optimizer to reorder instructions before this. On x64, it won't emit any memory barrier (as the architecture is TSO).

So, in this kind of code (that should be similar to a portion of `_isatty`):
```
int* pint = &global_int;
atomic_thread_fence(std::memory_order_acq_rel);
ioinfo **pio = &__pioinfo;
*pint = 42; 
pio->field = SOME_VALUE;
return;
```

Introducing the barrier ensure that pioinfo address is *always* loaded after global_int, with any optimization. So in this case, we'll keep the pattern as expected (latest `adrp/add` before `ret`).

**Comment by pbo-linaro** [comment]
Thanks for your collaboration on this :)

**Comment by macumber** [comment]
I'm on a new Windows ARM Dev Kit machine with Windows 11 23H2 22631.3085.  I've installed rubyinstaller-3.2.3-1-x64.exe to this machine.  I've downloaded and installed Windows Kits 10.0.25357.0 and 10.0.26020.0.  I thought that by adding the redistributable path to the front of my path, e.g. `$env:Path = 'C:\Program Files (x86)\Windows Kits\10\Redist\10.0.25357.0\ucrt\DLLs\x64\;' + $env:Path` that ruby.exe would find and load these new dlls.  However, I still get the following error :

```
PS C:\Ruby32-x64\bin> .\ruby.exe -e "puts 'hi'"
unexpected ucrtbase.dll
```

Is there some other step that I am missing?
